rep-0151.rst

```diff
@@ -0,0 +1,151 @@
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 23-Dec-2017
+
+Outline
+=======
+
+#. Abstract_
+#. Motivation_
+#. Rationale_
+#. Proposal_
+#. Requirements_
+#. References_
+#. Copyright_
+
+Abstract
+========
+
+This REP describes the path for switching from Python 2 to Python 3 in ROS 1.
+
+Motivation
+==========
+
+As described in the Python release schedule [1]_ Python 2.7 will only receive
+bugfix releases until 2020.
+As a consequence Ubuntu is trying to demote Python 2.7 into the `universe`
+repository for 18.04 since they don't want to guarantee support for 5 years.
+While it is unlikely that Python 2.7 will be removed from the archives entirely
+it should not be relied on past its EOL data.
+Therefore ROS 1 needs to begin moving to Python 3 in order to make the
+transition as smooth as possible.
+
+Rationale
+=========
+
+The goal for a future ROS 1 distribution to support Python 3 should be achieved
+no later than for ROS O-turtle (Mai 2020).
+Since O-turtle is a LTS release it is strongly desired to already support
+Python 3 in the ROS N-turtle release to allow enough testing time before the
+mandatory switch.
+
+In general a single Python version must be used since different versions can't
+be mixed within a single ROS distribution.
+E.g. a package A using Python 3 can't use the Python modules from package B if
+that was built with/for Python 2.
+
+In order to support Python 3 it is mandatory that the testing infrastructure
+(namely `devel` and `PR` jobs) can run the build and tests with Python 3.
+Since each of these jobs require Debian packages of the dependencies to be
+present being able to test for a specific Python version requires Debian
+packages for this Python version to be present.
+The alternative of building all dependencies from-source in these jobs doesn't
+scale and is therefore not feasible since the time necessary for a single build
+gets increasingly high the further up in the stack the package is.
+
+Updating all ROS packages which contain Python code to support Python 2
+as well as 3 is a significant effort.
+Therefore it is expected that this will take a while and will require the ROS
+community to share the necessary workload.
+
+Proposal
+========
+
+For a smooth transition the following timeline is being proposed:
+
+ * ROS O-turtle (May 2020, LTS):
+
+   * The only Debian packages provided are using Python 3.
+   * From-source builds are using Python 3 by default.
+   * All ROS packages should support Python 3.
+   * Python 2 support is not required anymore.
+
+ * ROS N-turtle (May 2019):
+
+   * The set of recommended Debian packages is using Python 3.
+   * There is an alternative set of Debian packages which is using Python 2,
+     this can be used in case some packages don't support Python 3 yet.
+   * From-source builds are using Python 3 by default, optionally choosing
+     Python 2 should be possible for all packages.
+   * All ROS packages should support Python 2 as well as Python 3.
```

Comment:
Again, "All actively-maintained ROS packages".

Reply:
IMO there needs to be a low-effort option here for leaf packages to support only one or the other. Maybe a flag in the `<export>` section that indicates "I've been ported to Python 3; don't try to build/test me for Python 2 any more."

Reply:
I don't think that we can expect the ROS community to do the switch within a single release. Therefore the requirement for dual support at least during the transition period makes sense imo.

Reply:
Right now you're not only insisting on doing the port in a single release, you're insisting that that the entire community port from Py2 -> Py2+3, which IMO is much more difficult than just porting from Py2 -> Py3. For non-module, non-core packages, there should be an option to do the port without having to be bilingual for a release or two.

Reply:
> Right now you're not only insisting on doing the port in a single release

That is simply not correct. The proposal explicitly gives maintainers time over a couple of releases to make the transition since I don't think the transition of every package in a single ROS release is feasible.

> you're insisting that that the entire community port from Py2 -> Py2+3, which IMO is much more difficult than just porting from Py2 -> Py3. For non-module, non-core packages, there should be an option to do the port without having to be bilingual for a release or two.

The technical fact is that a set of Debian packages *must* use the same Python version. You can not mix-and-match packages which only support different Python versions. That is also the reason why marking packages as Python 2 or Python 3 only in the manifest isn't feasible.

Yes, supporting both Python versions is more effort than just migrating to Python 3. Due to the technical limitation that we can't mix Python versions in a single distro and as you mentioned yourself that we don't want to force every package to move from 2 to 3 in a single release I propose the in-between state which support both. Imo the little extra effort to support both versions is out weighted by the benefit of stretching the transition over multiple releases.

If you don't want to support both version in your packages you could switch between M-turtle (using Python 2 by default) and N-turtle (using Python 3 by default). That would still make all "default" Debian packages work. The consequence is that all dependencies of your packages have to follow your lead. Also the "alternative" Debian packages wouldn't work since your package doesn't support the non-default Python version. It should be possible to disable your packages in the buildfarm configuration for the non-default Python version which will effectively remove your packages as well as all downstream packages from the set of alternative Debian packages. While not desired that is certainly doable.

Reply:
Perhaps there are some fairly common simple cases that could be made to work in Melodic with Python 3. 

Some packages only export scripts or nodes, which could select the Python version via the ``#!`` line. As long as there are no exported modules (though perhaps some internally used ones), the only constraint would be depending on Python 3 packages for the ROS core and other pure Python packages. That would probably require some kind of rosdep support, as already envisioned.

Perhaps there are other issues I am forgetting, right now.

Reply:
I don't understand how the proposal as written allows any flexibility at all for package authors. If I'm maintaining `my_fancy_package` currently, and want to release it for ROS Melodic, I can go ahead and bloom the thing, but if I haven't done the work _by then_ to make it bilingual, I will be immediately spammed with buildfarm emails about the unit tests failing, or compiled modules not building at all— as written, releasing to Melodic is a hard deadline.

My proposal is that a package maintainer be able to opt-out of the "alternative" package being built _at all_, as well as opting out of CI coverage for the Python version that they aren't (yet/any more) supporting. This will indeed prevent dependencies of such packages from being released, but many of the most-affected Python packages are leafs anyway. As long as all the "core" packages are in the alternative set, and the main set is complete, this does indeed give authors flexibility— they can choose to go bilingual in time for Melodic, or during Melodic, or in time for N Turtle, or if they're a non-module Python package, they can opt never to participate in the "alternative" scheme and just port to Python 3 exclusively in time for N Turtle, using the alternative packages of their dependencies to prepare and test in advance (eg on another branch).

The feasibility of all this depends a bit on how it gets implemented on the backend, hence my push for more implementation details. Like, will the rosdebian bloom plugin essentially double-bloom everything for the alternative releases, so that I end up with GBP tags for `ros-melodic-my-fancy-package` and `ros-melodicpy3-my-fancy-package`? Or will I need to manually bloom twice? Or will it still just be a single bloom and some trickery at the buildfarm layer where a separate "bin" job that produces the alternative deb by patching the GBP?

Reply:
> My proposal is that a package maintainer be able to opt-out of the "alternative" package being built at all, as well as opting out of CI coverage for the Python version that they aren't (yet/any more) supporting.

That would currently already be possible using the blacklisting in the buildfarm configuration.

Adding a flag to the manifest as an alternative location to configure this would be feasible but would require that functionality to be developed. Currently that is not planned. Maybe someone would like to contribute this?

> Like, will the rosdebian bloom plugin essentially double-bloom everything for the alternative releases, so that I end up with GBP tags for ros-melodic-my-fancy-package and ros-melodicpy3-my-fancy-package? Or will I need to manually bloom twice? Or will it still just be a single bloom and some trickery at the buildfarm layer where a separate "bin" job that produces the alternative deb by patching the GBP?

A single "release" in the gbp is not sufficient since the Debian control files need to be different between the Python version. We certainly don't want to require users to run `bloom` twice. So this will need to be handled internally by bloom / the buildfarm.

Reply:
So rather than packages having an explicit annotation for their supported Python versions, it will basically be a matter of wiki documentation and user discovery. Like, does my dependency work on Py3? Who knows; let's install it and see what goes bang...